### PR TITLE
NodePort Optimization: Filter out nodes based on nodeLabelSelector present in values.yaml

### PR DESCRIPTION
**Pull request by akshayhavile**
…

### CODE REVIEW

File: internal/k8s/controller.go [hunk 1]
```diff
@@ -947,6 +960,12 @@ func (c *AviController) SetupEventHandlers(k8sinfo K8sinformers) {
 			oldobj := old.(*corev1.Node)
 			node := cur.(*corev1.Node)
 			key := utils.NodeObj + "/" + node.Name
+			if lib.IsNodePortMode() {
+				if !lib.IsValidLabelOnNode(node.GetLabels(), key) {
```
Comment (monotosh-avi) [inline internal/k8s/controller.go:964]: i think we need to check the labels on the old obj too and process it if the old obj had the labels, else that node won't be removed from pool servers.

Comment (akshayhavile) [reply internal/k8s/controller.go]: addressed

File: internal/objects/nodes.go [hunk 2]
```diff
@@ -38,8 +39,18 @@ func SharedNodeLister() *K8sNodeStore {
 	return &K8sNodeStore{nodesStoreInstance}
 }
 
-func (o *K8sNodeStore) PopulateAllNodes(cs *kubernetes.Clientset) {
-	allNodes, _ := cs.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+func (o *K8sNodeStore) PopulateAllNodes(cs *kubernetes.Clientset, isNodePort bool, nodeLabels map[string]string) {
+	lOpt := metav1.ListOptions{}
```
[the remark this answers is not in the document]
Comment (akshayhavile) [reply internal/objects/nodes.go]: addressed

**Review by chauhanshubham — APPROVED** [review]

File: internal/k8s/controller.go [hunk 3]
```diff
@@ -947,6 +960,12 @@ func (c *AviController) SetupEventHandlers(k8sinfo K8sinformers) {
 			oldobj := old.(*corev1.Node)
 			node := cur.(*corev1.Node)
 			key := utils.NodeObj + "/" + node.Name
+			if lib.IsNodePortMode() {
+				if !lib.IsValidLabelOnNode(node.GetLabels(), key) && !lib.IsValidLabelOnNode(oldobj.GetLabels(), key) {
```
[the remark this answers is not in the document]
Comment (akshayhavile) [reply internal/k8s/controller.go:964]: yes. To remove servers from pool list, we require this check.

File: internal/objects/nodes.go [hunk 4]
```diff
@@ -38,8 +39,18 @@ func SharedNodeLister() *K8sNodeStore {
 	return &K8sNodeStore{nodesStoreInstance}
 }
 
-func (o *K8sNodeStore) PopulateAllNodes(cs *kubernetes.Clientset) {
-	allNodes, _ := cs.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+func (o *K8sNodeStore) PopulateAllNodes(cs *kubernetes.Clientset, isNodePort bool, nodeLabels map[string]string) {
+	lableOption := metav1.ListOptions{}
```
[the remark this answers is not in the document]
Comment (akshayhavile) [reply internal/objects/nodes.go]: addressed

**Comment by akshayhavile** [comment]
build ako

**Comment by akshayhavile** [comment]
build ako

**Review by chauhanshubham — APPROVED** [review]

**Review by swathinsankaran — APPROVED** [review]

**Pull request merged**